File: day01/solve.rs
use std::io::Read;

fn parse_list(list: String) -> Vec<Vec<i32>> {
    list.split("\n\n").map(|inventory| {
        inventory.lines()
            .map(|item| item.parse::<i32>().unwrap())
            .collect()
    }).collect()
}

fn sum(inventory: &Vec<i32>) -> i32 {
    inventory.iter().sum()
}

fn sums(inventories: Vec<Vec<i32>>) -> Vec<i32> {
    inventories.iter().map(sum).collect()
}

fn max(calories: Vec<i32>) -> i32 {
    *calories.iter().max().unwrap()
}

fn part1() {
    let input = read_input();
    print!("{}\n", max(sums(parse_list(input))));
}

fn part2() {
    let input = read_input();
    let mut total_per_elf = sums(parse_list(input));
    total_per_elf.sort_by(|a, b| b.cmp(a));
    print!("{}\n", sum(&total_per_elf[0..3].to_vec()));
}

fn read_input() -> String {
    let mut input = String::new();
    std::io::stdin().read_to_string(&mut input).unwrap();
    input
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    if args[1] == "1" {
        part1();
    } else if args[1] == "2" {
        part2();
    } else {
        panic!("invalid argument");
    }
}

File: day01/solve.py
#!/usr/bin/env python

import sys


def parse_inventories(inventories: str) -> list[list[int]]:
    """
    returns a list of list of ints

    e.g. input:  '1\n\n1\n2\n\n3\n\n4\n5\n6'
         output: [[1], [1, 2], [3], [4, 5, 6]]

    >>> parse_inventories("1\\n\\n1\\n2\\n\\n3\\n\\n4\\n5\\n6")
    [[1], [1, 2], [3], [4, 5, 6]]
    """
    return list(
        list(map(int, inventory.split("\n")))       # map to int
        for inventory in inventories.split("\n\n")  # split on double lines
    )


def part_one(data: str) -> int:
    """
    return the largest sum 
    """
    return max(sum(inventory) for inventory in parse_inventories(data))


def part_two(data: str) -> int:
    """
    return the sum of the 3 largest sums
    """
    return sum(sorted((sum(inventory) for inventory in parse_inventories(data)), reverse=True)[:3])


def read_input() -> str:
    """
    reads and returns the input from stdin
    """
    return sys.stdin.read().strip("\n")


def main(args):
    data = read_input()

    match a := args[0]:
        case "1":
            print(f"part_one: {part_one(data)}")
        case "2":
            print(f"part_two: {part_two(data)}")
        case _:
            raise ValueError(f"invalid argument '{a}'") 


if __name__ == "__main__":
    main(sys.argv[1:])

